### ISSUE TITLE
Add alt text label similar to twitter to encourage authors to make images accessible

### ISSUE THREAD

**Issue by domoritz** [mastodon/mastodon#18394]
### Pitch

Twitter recently enabled a small [ALT label in all images](https://help.twitter.com/en/using-twitter/add-image-descriptions) with alt texts and it greatly improved the visibility of alt texts. Mastodon should add this feature as well. 

### Motivation

This would benefit everyone. People who need alt texts will get more. Authors will know their alt texts are seen more.

**Comment by ellieayla** [comment]
https://docs.joinmastodon.org/user/posting/#attachments

https://github.com/mastodon/mastodon/issues/117 & https://github.com/mastodon/mastodon/pull/5123

**Comment by domoritz** [comment]
This is now what this request is about ;-) I want the little label in the corner to indicate that there is an alt text. Right now, I don't think there is a quick way to see whether an image has a description or not. 

<img width="128" alt="Screen Shot 2022-05-10 at 10 52 51 PM" src="https://user-images.githubusercontent.com/589034/167759356-65efcc93-15e4-4770-b225-be7ed45b9ded.png">

**Comment by mkody** [comment]
This would be a good feature to implement. Not to just show the "ALT" label but also for clicking on it so it can open a small popin/window to read the alt text. 

As a person who doesn't use accessibility tools, the only way I can read them is by hovering the image. Which shows the text in a very small tooltip that might cut off if the text is too long.

**Comment by ChartreuseK** [comment]
With the current hover only approach there's no easy way to copy paste the text, and you are limited to the tool-tip size, being able to pop it out to read it, or copy from it would be a nice improvement. 

As an alternative, perhaps the alt text could be shown below the image when an image is clicked on for larger viewing?

**Comment by domoritz** [comment]
I don't think the indicator should only appear on hover. It really helps to see that images have alt text. 

Take for example this tweet of mine with multiple pictures. I can tap on them to get the description but I can immediately see which ones have an alt text. 

![image](https://user-images.githubusercontent.com/589034/168207850-b47c714f-ab1f-4cf8-8dd7-2150a32d9a90.jpeg)

**Comment by koyuawsmbrtn** [comment]
Please add this to the next Mastodon release. We need this!

**Comment by aardrian** [comment]
If/when this feature gets built, I wrote up issues with Twitter's implementation that might be instructive:
_[Keyboard Challenges for Twitter’s New ALT Badge](https://adrianroselli.com/2022/04/keyboard-challenges-for-twitters-new-alt-badge.html)_

It goes beyond just keyboard, though, since I also point out ARIA, WHCM, and other interaction issues.

**Comment by Helianthropy** [comment]
Does hover work on touch enabled screens? I honestly can’t tell since there’s no guarantee that alt text exists on the images I’m trying. I also appreciate the reminder to users that alt text is a thing.

**Comment by aardrian** [comment]
> Does hover work on touch enabled screens?

My Surface laptop is a touch-enabled screen and I use a mouse with it. My Android phone can also be paired with a mouse. That being said, hover alone can never be a sufficient trigger.

**Comment by mgifford** [comment]
Authors need support to create accessible content. Here are some recommendations from a study for the European Commission https://accessibilitycluster.com/main-results/text-alternative-feature-%E2%80%93-presentation

**Comment by mkody** [comment]
> Does hover work on touch enabled screens?

They should appear if you long-press an image. Sometimes they appear truncated but tapping on the text again shows the full version.

<details>
<summary><b>Screenshots for example here</b></summary>

Android with Chrome/Vivaldi:
![image](https://user-images.githubusercontent.com/843787/201321749-4a72cf1c-1669-4503-9ef0-641726ac78cc.png)

Android with Firefox:
![image](https://user-images.githubusercontent.com/843787/201321795-e399a76a-b3c2-46e2-9aff-2763079f4204.png)

iPad with Safari:
![image](https://user-images.githubusercontent.com/843787/201321826-114e8ba6-7a02-4fc0-9776-afb85322e4c6.png)

</details>

**Comment by ellieayla** [comment]
Never got alt text to appear in iPhone Safari like it does in that iPad screenshot. 

<details>
  <summary>iPhone 11 running iOS 15.6.1 Screenshots</summary>

https://accessibility.huit.harvard.edu/describe-content-images long-tap on image

Safari on iPhone.

![33360AB2-2C56-435A-8A25-2AD494483F66](https://user-images.githubusercontent.com/1447600/201426980-fd7cba6a-1a57-4ddd-9e70-40a3de90b487.png)

compare with Chrome on iPhone:

![628F4586-E100-4C26-9BC0-231387CB8B5A](https://user-images.githubusercontent.com/1447600/201427006-14e8bb28-b55b-43c1-b620-6d2b0bc28861.png)
</details>

**Comment by mkody** [comment]
Oh just noticed that it doesn't work directly from a post on my side too... I need to expand the image first.

<details>
<summary><b>Video demo</b></summary>

https://user-images.githubusercontent.com/843787/201430276-04cd5055-5c5d-4f90-8f55-c4db95b3f6e2.mp4

</details>

**Comment by aardrian** [comment]
@mkody Note that the alt text is truncated. The entirety of it is not visible.

**Comment by beadsland** [comment]
Throwing my two cents into the well for a solution on this.

Apparently, on some touchscreen devices, some apps will provide ALT-text in the long-press contextual menu. e.g., am told that Chrome for Android does this. OTOH, my ChromeOS notebook does not.

As really don't want to boost un-ALTed images, have begun using Stylus to munge the CSS on those that do not have ALT (similar to [rkingett](https://github.com/rkingett)'s suggestion). This requires a certain level of technical knowledge to do, however, so isn't a solution for everyone.

Relatedly, per [ChartreuseK](https://github.com/ChartreuseK)'s comment, sometimes it can be useful to copy ALT text. For instance, I post various charts on a weekly basis, where the ALT-text is largely boilerplate, with slight edits from week to week. Being able to copy the ALT from the prior week to then mock it up for the next post a week later would be quite helpful. Yeah, this is a niche use case, and I could just maintain the updated boilerplate in another app outside mastodon (with all the extra clicks and screen switches that would entail), but am sure others could come up with similar scenarios where being able to copy ALT (without diving into an inspect window) would be helpful.

Further following up on ChartreuseK's ideas, would love it if there were simply a toggle to show/hide ALT text right below the toot of the image, much like CWs show/hide toot body. The tooltip text is quite small on hover compared to UI text of the rest of the interface, and hard to read due to lack of word wrapping. Even those of us who aren't using screen readers can benefit from being able to peruse ALT-text. Sometimes there's insights about the image that the author saw but aren't immediately apparent to the viewer. Sometimes it's just helpful for folk trying to learn how to write good ALT text to be able to read how others are doing it.

**Comment by mgifford** [comment]
Related issue here https://github.com/mastodon/mastodon/issues/19974

**Comment by ineffyble** [comment]
Merging in #22049:

# Pitch
I believe it would be beneficial to show the alt text next to the image when expanded, and have an "ALT" button that shows the alt text, similar to how Twitter does it (I think, haven't checked in a while).

# Motivation
Currently, for alt text, I have to hover over the image to view it. I didn't know this feature was there, I assumed it was a client-specific thing, as I've only seen the alt text on Tusky until now.

**Comment by ronilaukkarinen** [comment]
I have it like this:

![image](https://user-images.githubusercontent.com/1534150/229813772-00f6b58b-3966-4b1b-81f8-a30f2940c857.png)

It's rather easy to implement, [here](https://github.com/ronilaukkarinen/mastodon-bird-ui/blob/bdb167422e3064afdcee90cb57f8a39f4e82ed1c/style.css#L2953-L2986).

```css
.layout-single-column .media-gallery__item-thumbnail:has(> img:not([alt]))::after,
.layout-single-column .video-player:has(> video:not([title]))::after,
.layout-single-column .media-gallery__gifv:has(> video:not([title]))::after {
  align-self: flex-end;
  background-color: var(--color-bg-75);
  border-radius: 4px;
  bottom: 4px;
  color: var(--color-light-text);
  content: 'No alt';
  display: flex;
  font-size: 9px;
  font-weight: 500;
  height: 14px;
  justify-self: flex-end;
  left: auto;
  line-height: 14px;
  opacity: 1;
  padding: 2px 5px;
  position: absolute;
  right: 4px;
  text-transform: uppercase;
  top: auto;
  width: unset;
  z-index: 3;
}
```

**Comment by trwnh** [comment]
Fixed by #24782
